File: tests/rust/game/big_id.rs
//! The u64 fidelity pin: a bit-63 entity id (0x8000000000000001 — its
//! decimal exceeds i64) must survive create → query → format EXACTLY.
//! Rust carries ids as u64 natively, so the risk isn't a VM number type
//! — it's a careless float or i64 hop in the wrapper/parse path. Every
//! id below moves through pure u64 arithmetic; the asserts panic (→
//! eviction) on any drift, and the components never land.

use crate::labelle::{self, EntityId, Script};

const BIG: EntityId = 0x8000000000000001;

#[derive(Default)]
pub struct BigId {
    ids: Vec<EntityId>,
    scratch: Vec<u8>,
}

impl Script for BigId {
    fn init(&mut self) {
        // The Zig test pre-seeds the mock's next id to BIG.
        let id = labelle::create_entity();
        assert_eq!(id, BIG, "created id drifted");
        assert!(labelle::set_component(id, "Marker", r#"{"tag":42}"#));

        // Round-trip through the query path: the id crosses host JSON
        // and the wrapper's parse — bit-exactness is the whole point.
        assert!(labelle::query_into(
            r#"["Marker"]"#,
            &mut self.ids,
            &mut self.scratch
        ));
        assert_eq!(self.ids.len(), 1, "query missed the entity");
        assert_eq!(self.ids[0], BIG, "queried id drifted");

        // Write through the QUERIED id and render it unsigned — the Zig
        // side pins both the addressed entity and the exact decimal.
        let queried = self.ids[0];
        assert!(labelle::set_component(
            queried,
            "BigId",
            &format!("{{\"idstr\":\"{}\"}}", queried),
        ));
    }
}

File: tests/go/game/bigid.go
// The 64-bit id pin: go has a REAL uint64, so a bit-63 id must survive
// create → set → query → render with no signed detour and no float hop
// anywhere (the wrapper's parseIDs is pure uint64 arithmetic).
package game

import (
	"fmt"
	"strconv"

	"labelle"
)

type BigID struct{ labelle.NoopScript }

func (BigID) Init() {
	e := labelle.CreateEntity()
	if e != 0x8000000000000001 {
		panic("expected the forced bit-63 id")
	}
	labelle.SetComponent(e, "Marker", `{"tag":42}`)

	ids, ok := labelle.Query(`["Marker"]`)
	if !ok || len(ids) != 1 || ids[0] != e {
		panic("query did not round-trip the bit-63 id")
	}

	// The id renders unsigned end to end (9223372036854775809, not a
	// negative int64 or a rounded float).
	labelle.SetComponent(e, "BigId",
		fmt.Sprintf(`{"idstr":"%s"}`, strconv.FormatUint(ids[0], 10)))
}
